replica/table.cc

```diff
@@ -990,15 +996,21 @@ table::do_add_sstable_and_update_cache(sstables::shared_sstable sst, sstables::o
     co_return co_await get_row_cache().invalidate(row_cache::external_updater([&] () noexcept {
         // FIXME: this is not really noexcept, but we need to provide strong exception guarantees.
         // atomically load all opened sstables into column family.
-        compaction_group& cg = compaction_group_for_sstable(sst);
+        compaction_group* cg;
+        try {
+            cg = &compaction_group_for_sstable(sst);
```

Comment:
how can the compaction group be gone when we get here?

Reply:
maybe a compaction_group needs to be allocated?
the calling paths for add_sstable_and_update_cache are: table_populator::populate_subdir, streaming, and unit tests.

And for add_sstables_and_update_cache: distributed_loader::make_sstables_available (during initial population).
For that path, does it mean that we didn't populate system.tablets yet, or maybe those are stale sstables left over after tablets got migrated?
Anyway, we can't just ignore this error, it must be handled with care.

Reply:
This looks like a race between repair and tablet migration
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7274/artifact/testlog/x86_64/debug/scylla-1957.log
```
Scylla version 5.5.0~dev-0.20240311.15fc04c01cfb with build-id 7254dd2cca5c13154275141c509811bace1f394b starting ...
...
INFO  2024-03-11 15:16:10,498 [shard 0:strm] repair - repair[3eb90a13-ae80-4f72-901d-05b5294ae5d4]: Started to repair 1 out of 1 tables in keyspace=test, table=test, table_id=869ff100-dfa9-11ee-b652-d1d590cee262, repair_reason=repair
INFO  2024-03-11 15:16:10,511 [shard 1:strm] table - Cleaned up tablet 28 of table test.test successfully.
...
terminate called after throwing an instance of 'std::runtime_error'
  what():  storage_group_for_token: tablet is not allocated in this shard: id=28, token=7087100466026478816
```

Decoded:
```
replica::tablet_storage_group_manager::compaction_group_for_sstable(seastar::lw_shared_ptr<sstables::sstable> const&) const at table.cc:?
replica::table::compaction_group_for_sstable(seastar::lw_shared_ptr<sstables::sstable> const&) const at ./replica/table.cc:944
 (inlined by) operator() at ./replica/table.cc:999
 (inlined by) seastar::noncopyable_function<void ()>::direct_vtable_for<replica::table::do_add_sstable_and_update_cache(seastar::lw_shared_ptr<sstables::sstable>, seastar::bool_class<sstables::offstrategy_tag>, bool)::$_0>::call(seastar::noncopyable_function<void ()> const*) at ././seastar/include/seastar/util/noncopyable_function.hh:129
row_cache::external_updater::execute() at ././row_cache.hh:209
 (inlined by) operator() at ./row_cache.cc:1445
 (inlined by) seastar::future<void> seastar::futurize<seastar::future<void> >::invoke<row_cache::do_update(row_cache::external_updater, std::function<seastar::future<void> ()>)::$_0::operator()<row_cache::external_updater, std::function<seastar::future<void> ()> >(row_cache::external_updater&, std::function<seastar::future<void> ()>&) const::{lambda(auto:1)#1}::operator()<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >(seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>)::{lambda()#1}&>(row_cache::external_updater&&) at ././seastar/include/seastar/core/future.hh:2032
 (inlined by) auto seastar::futurize_invoke<row_cache::do_update(row_cache::external_updater, std::function<seastar::future<void> ()>)::$_0::operator()<row_cache::external_updater, std::function<seastar::future<void> ()> >(row_cache::external_updater&, std::function<seastar::future<void> ()>&) const::{lambda(auto:1)#1}::operator()<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >(seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>)::{lambda()#1}&>(row_cache::external_updater&&) at ././seastar/include/seastar/core/future.hh:2063
```